File: app/src/main/java/com/lin/read/adapter/SearchBookItemAdapter.java
package com.lin.read.adapter;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.lin.read.R;
import com.lin.read.filter.BookInfo;

import java.util.ArrayList;

/**
 * Created by lisonglin on 2017/10/17.
 */

public class SearchBookItemAdapter extends RecyclerView.Adapter<SearchBookItemAdapter.ViewHolder> {
    private ArrayList<BookInfo> allBookData;
    private Context context;

    public SearchBookItemAdapter(Context context, ArrayList<BookInfo> allBookData) {
        this.allBookData = allBookData;
        this.context = context;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.item_search_book, null);
        ViewHolder holder = new ViewHolder(view);
        return holder;
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        final BookInfo bookInfo=allBookData.get(position);
        holder.bookName.setText(bookInfo.getBookName());
        holder.authorName.setText(bookInfo.getAuthorName());
        holder.bookType.setText(bookInfo.getBookType());
        holder.lastUpdate.setText(bookInfo.getLastUpdate());
        holder.lastContent.setText(bookInfo.getLastChapter());
        holder.download.setEnabled(!TextUtils.isEmpty(bookInfo.getDownloadLink()));

        holder.download.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if(TextUtils.isEmpty(bookInfo.getDownloadLink())){
                    Toast.makeText(context,"未找到下载资源",Toast.LENGTH_SHORT).show();
                }else{
                    Intent intent = new Intent();
                    intent.setAction("android.intent.action.VIEW");
                    Uri content_url = Uri.parse(bookInfo.getDownloadLink());
                    intent.setData(content_url);
                    context.startActivity(intent);
                }
            }
        });
    }

    @Override
    public int getItemCount() {
        return allBookData==null?0:allBookData.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder{

        private TextView bookName;
        private TextView authorName;
        private TextView bookType;
        private TextView lastUpdate;
        private TextView lastContent;
        private Button download;
        public ViewHolder(View itemView) {
            super(itemView);
            bookName= (TextView) itemView.findViewById(R.id.book_item_bookname);
            authorName= (TextView) itemView.findViewById(R.id.book_item_authorname);
            bookType= (TextView) itemView.findViewById(R.id.book_item_booktype);
            lastUpdate= (TextView) itemView.findViewById(R.id.book_item_lastupdate);
            lastContent= (TextView) itemView.findViewById(R.id.book_item_lastcontent);
            download= (Button) itemView.findViewById(R.id.book_item_download);
        }
    }
}

File: app/src/main/java/com/lin/read/adapter/SearchBookItemAdapter.kt
package com.lin.read.adapter

import android.content.Context
import android.content.Intent
import android.os.Parcelable
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.lin.read.R
import com.lin.read.activity.ReadBookActivity
import com.lin.read.filter.ReadBookBean
import com.lin.read.filter.SearchBookBean
import com.lin.read.utils.Constants
import kotlinx.android.synthetic.main.item_search_book.view.*

class SearchBookItemAdapter(private val context: Context, private val allBookData: List<SearchBookBean>) : RecyclerView.Adapter<SearchBookItemAdapter.ViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): ViewHolder {
        return ViewHolder(LayoutInflater.from(context).inflate(R.layout.item_search_book, null).apply {
            setOnClickListener{
                readBook(allBookData[it.tag as Int])
            }
        })
    }

    override fun getItemCount(): Int {
        return allBookData.size
    }

    override fun onBindViewHolder(holder: ViewHolder?, position: Int) {
        holder?.view?.tag = position
        allBookData[position].run {
            holder?.itemView?.book_item_bookname?.text = bookName
            holder?.itemView?.book_item_authorname?.text = authorName
            holder?.itemView?.book_item_lastupdate?.text = lastUpdate
            holder?.itemView?.book_item_lastcontent?.text = lastChapter
            holder?.itemView?.book_item_read?.setOnClickListener {
                readBook(this)
            }
        }
    }

    private fun readBook(searchBookBean: SearchBookBean) {
        searchBookBean.run {
            val intent = Intent(context, ReadBookActivity::class.java)
            intent.putExtra(Constants.KEY_SKIP_TO_READ, ReadBookBean(webType, bookName, authorName, chapterLink) as Parcelable)
            context.startActivity(intent)
        }
    }

    class ViewHolder(val view: View) : RecyclerView.ViewHolder(view)
}